Editorial: Fix incorrect use of UnicodeMatchPropertyValue

gibson042 opened the pull request:
Fixes #3586
Closes #3590

Also includes commits with incidental fixes in nearby algorithms and steps.

gibson042 committed:
Editorial: Fix incorrect use of UnicodeMatchPropertyValue

Fixes #3586

gibson042 committed:
Editorial: Fix incorrect use of UnicodeMatchPropertyValue

Fixes #3586

gibson042 committed:
Editorial: Narrow an assertion about the output from UnicodeMatchProperty

gibson042 committed:
Editorial: Narrow an assertion about the output from UnicodeMatchProperty

gibson042 committed:
Editorial: Correct references to the structure of Unicode Character Databse file PropertyValueAliases.txt

gibson042 committed:
Editorial: Correct references to the structure of Unicode Character Databse file PropertyValueAliases.txt

Jack-Works commented on spec.html:
```diff
@@ -38578,7 +38579,7 @@ <h1>
             1. Assert: _v_ is a property value or property value alias for the Unicode property _p_ listed in <a href="https://unicode.org/Public/UCD/latest/ucd/PropertyValueAliases.txt"><code>PropertyValueAliases.txt</code></a>.
-            1. Let _value_ be the canonical property value of _v_ as given in the “Canonical property value” column of the corresponding row.
+            1. Let _value_ be the value in the “long name” field of the corresponding line in that file.
             1. Return the List of Unicode code points _value_.
```
Why does this AO return "a Unicode property value", but here "Return the List of ..."?

gibson042 replied:
Both UnicodeMatchProperty and UnicodeMatchPropertyValue are implicitly implementing each of "Unicode property name" and "Unicode property value" as a List of code points (similar to [ECMAScript source text](https://tc39.es/ecma262/multipage/ecmascript-language-source-code.html#sec-source-text) and the output of [StringToCodePoints](https://tc39.es/ecma262/multipage/ecmascript-language-source-code.html#sec-stringtocodepoints)), which I think is necessary for comparison against values in the [Unicode property tables](https://tc39.es/ecma262/multipage/text-processing.html#table-nonbinary-unicode-properties). There may be a better way, but this is is valid AFAICT.

michaelficarra replied:
I don't think it's invalid, but I also don't think it needs to be made explicit. IMO it can be returned directly. Though that should probably be a separate PR.

michaelficarra replied:
Actually I was pretty unhappy with the wording used in this PR to refer to external data or table values. I've pushed a commit to make it a little more consistent and explicit. This has also inspired me to add two new editorial conventions:

* assertion steps should not be necessary to understand an algorithm; a reader should interpret it the same if they are removed
* do not use determiners or pronouns to refer to values from other steps (e.g. "that number"); instead uses aliases or repetition

michaelficarra committed:
wording in UnicodeMatchProperty and UnicodeMatchPropertyValue

michaelficarra committed:
wording in UnicodeMatchProperty and UnicodeMatchPropertyValue

bakkot commented on spec.html:
```diff
@@ -38545,11 +38546,10 @@ <h1>
+            1. If _rer_.[[UnicodeSets]] is *true* and _p_ is listed in the “<emu-not-ref>Property name</emu-not-ref>” column of <emu-xref href="#table-binary-unicode-properties-of-strings"></emu-xref>, then
+              1. Return _p_.
+            1. Assert: _p_ is listed in the “<emu-not-ref>Property name</emu-not-ref> and aliases” column of <emu-xref href="#table-nonbinary-unicode-properties"></emu-xref> or <emu-xref href="#table-binary-unicode-properties"></emu-xref>.
+            1. Return the “canonical <emu-not-ref>property name</emu-not-ref>” corresponding to the <emu-not-ref>property name</emu-not-ref> or property alias _p_ in <emu-xref href="#table-nonbinary-unicode-properties"></emu-xref> or <emu-xref href="#table-binary-unicode-properties"></emu-xref>..
```
```suggestion
            1. Return the “canonical <emu-not-ref>property name</emu-not-ref>” corresponding to the <emu-not-ref>property name</emu-not-ref> or property alias _p_ in <emu-xref href="#table-nonbinary-unicode-properties"></emu-xref> or <emu-xref href="#table-binary-unicode-properties"></emu-xref>.
```

gibson042 commented on spec.html:
```diff
@@ -38578,8 +38578,7 @@ <h1>
             1. Assert: _v_ is a property value or property value alias for the Unicode property _p_ listed in <a href="https://unicode.org/Public/UCD/latest/ucd/PropertyValueAliases.txt"><code>PropertyValueAliases.txt</code></a>.
-            1. Let _value_ be the canonical property value of _v_ as given in the “Canonical property value” column of the corresponding row.
-            1. Return the List of Unicode code points _value_.
+            1. Return the “long name” corresponding to the “long name” or “short name” _v_ in <a href="https://unicode.org/Public/UCD/latest/ucd/PropertyValueAliases.txt"><code>PropertyValueAliases.txt</code></a>.
```
@michaelficarra I had to adjust this for correctness; _v_ might already be a long name here.

michaelficarra replied:
> Second Field: The second field is the short name for the property value.
> It is typically an abbreviation, but in a number of cases it is simply
> a duplicate of the "long name" in the third field.

This is unambiguously the "short name", even if it's also the "long name", no?

michaelficarra replied:
The problem I have with the wording now is that it permits a scenario where `v` is a long name *as well as* a short name associated with some other long name (even though that is likely prohibited in the Unicode data file). How about something like this?

```suggestion
            1. If _v_ is a “short name” associated with some “long name” _l_ for <emu-not-ref>property name</emu-not-ref> _p_ in <a href="https://unicode.org/Public/UCD/latest/ucd/PropertyValueAliases.txt"><code>PropertyValueAliases.txt</code></a>, return _l_; otherwise, return _v_.
```

gibson042 replied:
That doesn't quite work because there may be multiple aliases as in e.g. `/\p{Script=Qaac}/u` (<q><em>sc ; Copt                             ; Coptic                           ; Qaac</em></q> specifies short names "Copt" and "Qaac" for long name "Coptic"), but I see where you're going. Updated.

michaelficarra replied:
That new wording WFM.

Jack-Works commented:
I have another question, if you search "scx" in https://unicode.org/Public/UCD/latest/ucd/PropertyValueAliases.txt, you will find there is no record for it. And according to the [spec](https://tc39.es/ecma262/#sec-patterns-static-semantics-early-errors),

> [UnicodePropertyValueExpression](https://tc39.es/ecma262/#prod-UnicodePropertyValueExpression) :: [UnicodePropertyName](https://tc39.es/ecma262/#prod-UnicodePropertyName) = [UnicodePropertyValue](https://tc39.es/ecma262/#prod-UnicodePropertyValue)
> > It is a Syntax Error if the [source text matched by](https://tc39.es/ecma262/#sec-algorithm-conventions-syntax-directed-operations) [UnicodePropertyValue](https://tc39.es/ecma262/#prod-UnicodePropertyValue) is not a property value or property value alias for the Unicode property or property alias given by the [source text matched by](https://tc39.es/ecma262/#sec-algorithm-conventions-syntax-directed-operations) [UnicodePropertyName](https://tc39.es/ecma262/#prod-UnicodePropertyName) listed in [PropertyValueAliases.txt](https://unicode.org/Public/UCD/latest/ucd/PropertyValueAliases.txt).

All RegExp that have the form of `/\p{Script_Extensions=Anything}/u` (e.g., `/\p{Script_Extensions=Zanabazar_Square}/u`) are an early error, but it is not true in the implementation.

Jack-Works commented:
> I have another question, if you search "scx" in [unicode.org/Public/UCD/latest/ucd/PropertyValueAliases.txt](https://unicode.org/Public/UCD/latest/ucd/PropertyValueAliases.txt), you will find there is no record for it. And according to the [spec](https://tc39.es/ecma262/#sec-patterns-static-semantics-early-errors),
> 
> > [UnicodePropertyValueExpression](https://tc39.es/ecma262/#prod-UnicodePropertyValueExpression) :: [UnicodePropertyName](https://tc39.es/ecma262/#prod-UnicodePropertyName) = [UnicodePropertyValue](https://tc39.es/ecma262/#prod-UnicodePropertyValue)
> > > It is a Syntax Error if the [source text matched by](https://tc39.es/ecma262/#sec-algorithm-conventions-syntax-directed-operations) [UnicodePropertyValue](https://tc39.es/ecma262/#prod-UnicodePropertyValue) is not a property value or property value alias for the Unicode property or property alias given by the [source text matched by](https://tc39.es/ecma262/#sec-algorithm-conventions-syntax-directed-operations) [UnicodePropertyName](https://tc39.es/ecma262/#prod-UnicodePropertyName) listed in [PropertyValueAliases.txt](https://unicode.org/Public/UCD/latest/ucd/PropertyValueAliases.txt).
> 
> All RegExp that have the form of `/\p{Script_Extensions=Anything}/u` (e.g., `/\p{Script_Extensions=Zanabazar_Square}/u`) are an early error, but it is not true in the implementation.

Should this section add a reference to the <https://unicode.org/Public/UCD/latest/ucd/Scripts.txt>? I guess this is the source of the script list.

gibson042 committed:
Editorial: Fix incorrect use of Unicode property Script_Extensions

gibson042 committed:
Editorial: Fix incorrect use of Unicode property Script_Extensions

gibson042 committed:
Editorial: Add special cases for set-valued Unicode property Script_Extensions

cf. https://www.unicode.org/reports/tr18/#Script_Property and
https://www.unicode.org/reports/tr24/#Script_Extensions

gibson042 committed:
Editorial: Add special cases for set-valued Unicode property Script_Extensions

cf. https://www.unicode.org/reports/tr18/#Script_Property and
https://www.unicode.org/reports/tr24/#Script_Extensions

gibson042 commented:
Nice observation, @Jack-Works! Unicode property Script_Extensions (scx) is unusual in being set-valued rather than scalar-valued, and as such need special consideration in our spec. I have added editorial corrections to this PR, and opened #3590 for a potential followup.

Jack-Works commented on spec.html:
```diff
@@ -38282,19 +38285,24 @@ <h1>
+          1. If _p_ is `Script_Extensions`, then
+            1. Assert: _vs_ is a property value or property value alias for property “Script” listed in <a href="https://unicode.org/Public/UCD/latest/ucd/PropertyValueAliases.txt"><code>PropertyValueAliases.txt</code></a>.
+            1. Let _v_ be the Set containing the “short name”, “long name”, and any other aliases corresponding with value _vs_ for property “Script” in <a href="https://unicode.org/Public/UCD/latest/ucd/PropertyValueAliases.txt"><code>PropertyValueAliases.txt</code></a>.
+            1. Return the CharSet containing all Unicode code points whose character database definition includes the property “Script_Extensions” with value having a non-empty intersection with _v_.
```
Do you need to call `MaybeSimpleCaseFolding` here?

gibson042 replied:
Hmm, that would affect any code point that case-folds across script (or to/from Common). I don't know if there are any, but it's easy enough to accommodate. Done.

jmdyck commented on spec.html:
```diff
@@ -39340,12 +39348,11 @@ <h1>
             1. Assert: _v_ is a property value or property value alias for the Unicode property _p_ listed in <a href="https://unicode.org/Public/UCD/latest/ucd/PropertyValueAliases.txt"><code>PropertyValueAliases.txt</code></a>.
-            1. Let _value_ be the canonical property value of _v_ as given in the “Canonical property value” column of the corresponding row.
-            1. Return the List of Unicode code points _value_.
+            1. [declared="l"] If _v_ is a “short name” or other alias associated with some “long name” _l_ for <emu-not-ref>property name</emu-not-ref> _p_ in <a href="https://unicode.org/Public/UCD/latest/ucd/PropertyValueAliases.txt"><code>PropertyValueAliases.txt</code></a>, return _l_; otherwise, return _v_.
```
```suggestion
            1. [declared="l"] If _v_ is a “short name” or other alias associated with some “long name” _l_ for <emu-not-ref>property name</emu-not-ref> _p_ in <a href="https://unicode.org/Public/UCD/latest/ucd/PropertyValueAliases.txt"><code>PropertyValueAliases.txt</code></a>, return _l_; else return _v_.
```

gibson042 committed:
fixup! wording in UnicodeMatchProperty and UnicodeMatchPropertyValue
